fix: handle OOMs during KV cache estimation

ixlmar opened the pull request:
# fix: handle OOMs during KV cache estimation

## Description

Check for and report (OOM) errors encountered during model forward calls initiated by KV cache capacity estimation.

Changes also improve cleanup/shutdown behavior.

~~**Note:** Based on (and to be merged after) https://github.com/NVIDIA/TensorRT-LLM/pull/4581~~

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--disable-fail-fast --skip-test --stage-list "A10-1, xxx" --gpu-type "A30, H100_PCIe" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-[Post-Merge]-1, xxx"]`

Launch build/test pipelines. All previously running jobs will be killed.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests. Will also run L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-[Post-Merge]-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-[Post-Merge]-1, xxx".

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

ixlmar commented:
/bot run

tensorrt-cicd commented:
[PR_Github #6599](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/6599/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #6646](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/6646/) [ run ] triggered by Bot

ixlmar commented:
/bot reuse-pipeline

tensorrt-cicd commented:
[PR_Github #6746](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/6746/) [ reuse-pipeline ] completed with state `SUCCESS`
Reusing [PR_Github #6646](https://prod.blsm.nvidia.com/sw-tensorrt-top-1/job/LLM/job/helpers/job/PR_Github/6646/) for commit `6625c59`

ixlmar committed:
fix: handle OOMs during KV cache estimation

fix: cleanup executor loop on error (avoids hang)
fix: error check in estimate_max_kv_cache_tokens
fix: gracefully handle failed requests
Signed-off-by: ixlmar <206748156+ixlmar@users.noreply.github.com>

tensorrt-cicd commented:
[PR_Github #7077](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/7077/) [ run ] triggered by Bot

ixlmar commented:
/bot run

tensorrt-cicd commented:
[PR_Github #7314](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/7314/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #7314](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/7314/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #5302](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/5302/) completed with status: 'SUCCESS'
Pipeline passed with automatic retried tests. Check the [rerun report](https://urm.nvidia.com/artifactory/sw-tensorrt-generic/llm-artifacts//LLM/main/L0_MergeRequest_PR/5302/test-results/rerun_report.html) for details.

dcampora reviewed (approved):
LGTM

tensorrt-cicd commented:
[PR_Github #7486](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/7486/) [ run ] completed with state `FAILURE`
[/LLM/main/L0_MergeRequest_PR pipeline #5431](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/5431/) completed with status: 'FAILURE'

tensorrt-cicd commented:
[PR_Github #7493](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/7493/) [ run ] completed with state `ABORTED`

tensorrt-cicd commented:
[PR_Github #7507](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/7507/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #5444](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/5444/) completed with status: 'SUCCESS'
